docs/multiple_image_support_design_proposal.md

```diff
@@ -0,0 +1,549 @@
+uint64_t get_image_version(uint32_t image_id, uint32_t slot);
+
+i = 0;
+while (i < IMAGE_CNT)
+{
+    curr_slot = (boot_data[i].swap_type != NONE) ? 1 : 0;
+
+    ptr = get_dependency_entry(i, curr_slot);
+    while(ptr)
+    {
+        id = get_dependent_image_id(ptr);
+        min_version = get_dependent_image_min_version(ptr);
+        get_next_dependency_entry(&ptr);
+
+        dep_slot = (boot_data[id].swap_type != NONE) ? 1 : 0;
+        dep_version = get_image_version(id, dep_slot);
+
+        if (min_version <= dep_version)
+        {
+            /* Dependency satisfied, no action needed */
+            i++;
+        }
+        else
+        {
+            /* Dependency not satisfied */
+            switch (boot_data[i].swap_type)
+            {
+                case PERM:
+                case TEST:
+                    boot_data[i].swap_type = NONE;
+                break;
+                case NONE:
+                    boot_data[i].swap_type = REVERT;
+                break;
+            }
+
+            /* Dependency check needs to be restarted */
+            i = 0;
+            break;
+        }
+    }
+}
+```
+**
+
+Valid states after dependency check:
+  - In a consistent state images should move in the same direction or stay in
+    place. This means in practice that `swap_type` cannot be at the same time
+    `REVERT` and `<PERM | TEST>`.
+  - These `swap_type` combinations are valid:
+    - `NONE` or `<PERM | TEST>`
+    - `REVERT` or `NONE`
+
+# Testing in simulator
+Simulator environment must be improved to able to test possible scenarios during
+multiple image update, focusing on abort and revert scenarios and on the usage
+of common scratch area.
+
+# Testing on hardware
+Testing can be done on ARMv8-M based platform i.e.: MPS2_AN521 or Musca A1.
+MCUBoot can be compiled without knowing that it is running on ARMv8-M. Code,
+which was compiled to ARMv7-M is compatible with ARMv8-M. Just the memory map
+must be correct. Also the compiler does not need to support ARMv8-M
+architecture.
+
+Test application, specially the secure code must be compiled with ARMv8-M
+awareness. Currently this can be done in TF-M project.
+
+Testing on ARMv7-M based platform is also possible. Just need to compile two
+application with different and not overlapping memory map. They can be executed
+together. The first application can be started by the bootloader and second
+application is started by the first one, similar to secure and non-secure code
+on ARMv8-M.
+
+# Alternative solution
+Use only one secondary area and upgrade the images one-by-one after each other.
+Size of secondary area must be as big as the biggest image primary area.
+Benefits:
+  - Reduced the complexity of the code, because this is basically the currently
+    used process with single image update.
+  - Reduced flash fragmentation
+  - Better utilization of flash
+Drawbacks:
+  - When API is changing between images then the first image which is updated
```

Comment:
I'm not sure this is a reasonable restriction, and makes APIs difficult to maintain. I think we're best trying to find the full, correct, solution.

Reply:
I agree this solution makes harder to maintain APIs. I mentioned here just to cover each solution idea. It can be feasible if API is stable between images and there are strict storage space constrains.